pkg/xen-tools/patches-4.15.0/14-qemu-Init-CPU-mask-per-VCPU.patch

```diff
@@ -0,0 +1,181 @@
+From 658e687106d1a594fdcd54cb827f6989b1e74f53 Mon Sep 17 00:00:00 2001
+From: Nikolay Martyanov <ohmspectator@gmail.com>
+Date: Wed, 28 Sep 2022 15:52:24 +0200
+Subject: [PATCH 14/15] qemu: Init CPU mask per VCPU.
+
+Signed-off-by: Nikolay Martyanov <ohmspectator@gmail.com>
+---
+ tools/qemu-xen/include/hw/boards.h   |  1 +
+ tools/qemu-xen/include/hw/core/cpu.h |  2 +
+ tools/qemu-xen/softmmu/cpus.c        | 31 ++++++++++++
+ tools/qemu-xen/softmmu/vl.c          | 73 ++++++++++++++++++++++++++++
+ 4 files changed, 107 insertions(+)
+
+diff --git a/tools/qemu-xen/include/hw/boards.h b/tools/qemu-xen/include/hw/boards.h
+index b06f13e..a4b4f11 100644
+--- a/tools/qemu-xen/include/hw/boards.h
++++ b/tools/qemu-xen/include/hw/boards.h
+@@ -296,6 +296,7 @@ struct MachineState {
+     char *initrd_filename;
+     bool cpu_pin;
+     char *cpumask_str;
++    uint64_t cpumask;
+     const char *cpu_type;
+     AccelState *accelerator;
+     CPUArchIdList *possible_cpus;
+diff --git a/tools/qemu-xen/include/hw/core/cpu.h b/tools/qemu-xen/include/hw/core/cpu.h
+index 8f14573..628b31a 100644
+--- a/tools/qemu-xen/include/hw/core/cpu.h
++++ b/tools/qemu-xen/include/hw/core/cpu.h
+@@ -374,6 +374,8 @@ struct CPUState {
+     bool created;
+     bool stop;
+     bool stopped;
++    bool pinned;
++    uint64_t cpumask;
+     bool unplug;
+     bool crash_occurred;
+     bool exit_request;
+diff --git a/tools/qemu-xen/softmmu/cpus.c b/tools/qemu-xen/softmmu/cpus.c
+index a802e89..5a68859 100644
+--- a/tools/qemu-xen/softmmu/cpus.c
++++ b/tools/qemu-xen/softmmu/cpus.c
+@@ -2011,6 +2011,23 @@ static void qemu_dummy_start_vcpu(CPUState *cpu)
+                        QEMU_THREAD_JOINABLE);
+ }
+ 
++static inline void cpumask_clear_bit(uint64_t *mask, uint8_t bit)
++{
++    *mask &= ~(1ul << bit);
++}
++
++static inline long cpumask_get_min_bit(uint64_t mask)
++{
++  return __builtin_ffsll(mask) - 1;
++}
++
++static long pick_pcpu(uint64_t *cpumask)
++{
++    long ret = cpumask_get_min_bit(*cpumask);
++    cpumask_clear_bit(cpumask, ret);
++    return ret;
++}
++
+ void qemu_init_vcpu(CPUState *cpu)
+ {
+     MachineState *ms = MACHINE(qdev_get_machine());
+@@ -2019,6 +2036,20 @@ void qemu_init_vcpu(CPUState *cpu)
+     cpu->nr_threads =  ms->smp.threads;
+     cpu->stopped = true;
+     cpu->random_seed = qemu_guest_random_seed_thread_part1();
++    cpu->pinned = ms->cpu_pin;
++    static uint64_t vm_cpumask;
++    uint64_t vcpu_cpumask;
++    if (!vm_cpumask)
++        vm_cpumask = ms->cpumask;
++    if (!cpu->pinned) {
++        //If the CPUs are not pinned, assign the whole CPU mask to the VCPU
```

Comment:
I didn't find a checkpatch, but in the rest of the code comments are made with /*  */

Reply:
Will fix, thanks